src/book.py

```diff
@@ -560,22 +561,99 @@ def _read_kraken_ledgers(self, file_path: Path) -> None:
                 assert coin
                 assert change
 
-                self.append_operation(
-                    operation, utc_time, platform, change, coin, row, file_path
-                )
-
-                if fee != 0:
+                # append all operations to main list per default
+                # will be overwritten for deposits / withdrawals
+                append_operation = True
+
+                # Skip duplicate entries for deposits / withdrawals and additional
+                # deposit / withdrawal lines for staking / unstaking / staking reward
+                # actions.
+                # The second deposit and the first withdrawal need to be considered,
+                # since these are the points in time where the user actually has the
+                # assets at their disposal. The first deposit and second withdrawal are
+                # in the public trade history and are skipped.
+                # For staking / unstaking / staking reward actions, deposits /
+                # withdrawals only occur once and will be ignored.
+                if operation in ["Deposit", "Withdrawal"]:
+                    # search for refid in refids list
+                    refid_idxs = [
+                        idx for idx, op in enumerate(held_operations)
+                        if op["refid"] == refid
+                    ]
```

Comment:
`idx` is never used alone, but always combined with `held_operations[idx]`.

So this list could also be `matching_operations = [op ....]` or something like this.

```
if len(matching_operations) == 0:
    # Matching operation wasn't found
    matching_operation = None
elif len(matching_operations) == 1:
    matching_operation = matching_operations[0]
else:
    log.error(...)
    raise RuntimeError
```

btw, this might be faster, when we use a dictionary instead: `dict[ref_id, ...]`.

Reply:
For the dict approach we need to save something with at least three states (not seen id, seen id, second id). We could use bool for that (None, False, True).

When we see the first deposit, we set it to False. When we see the second deposit, we set it to True and add the operation.

When we see the first withdrawal, we add the operation and set False. When we see the second withdrawal we set True.

Afterwards, we check wether all values in the dictionary are true or raise a warning, that we are missing a specific refid.

Reply:
Hey @provinzio, I'm not sure if can follow you completely, I have some follow-up questions:

> `idx` is never used alone, but always combined with `held_operations[idx]`.
> So this list could also be `matching_operations = [op ....]` or something like this.

The "real operation item" op is only created afterwards when `append_operation` is called (not to be mixed with the `operation` variable, which is `"Deposit"` or `"Withdrawal"`). Therefore, `matching_operations` would need to be a subset of the `held_operations` list, but I don't see any advantage in that? Instead, we could introduce a temporary variable `matching_operation = held_operations[idx]` after checking the length of `idx` to prevent the repetition of `held_operations[idx]`?

> btw, this might be faster, when we use a dictionary instead: `dict[ref_id, ...]`. We can then use `held_operations.get(refid)` and check against None, wether we already saved something.

When you say that we could use a dictionary instead, do you mean a dictionary of dictionaries?:
```
held_operations = {
  {<refid1>: {'operation': <operation_name1>, 'utc_time': <utc_time1>, ...},
  {<refid2>: {'operation': <operation_name2>, 'utc_time': <utc_time2>, ...},
  ...
}
```

> As we do not want to save more than one operation, I think we do not need to save a list of operations in the dictionary, but just the relevant information like a string "first deposit found" or just True and for the withdrawal the previous operation (or add the withdrawal directly, but save it in the dict, to check if we found the second one.. might be better.

The `held_operations` contains the held-back operations of the whole CSV. It could be that between two deposits/withdrawals with the same refid there is another deposit/withdrawal action, so we need to keep track of all "open" held-back operations.
Additionally, we need to store all relevant information to append the operation at a later point in time (when the second withdrawal occurs, release/append the first withdrawal operation). This also allows us to compare the data of the first and second deposit/withdrawal.

> For the dict approach we need to save something with at least three states (not seen id, seen id, second id). We could use bool for that (None, False, True).

Do you want to store this as a separate variable (see below, all should be True at the end), or as part of `held_operations`?
```
refid_check = {
  <refid1>: None,
  <refid2>: False,
  <refid3>: True,
  ...
}
```

Reply:
Hey @Griffsano,


My comment on `matching_operations` doesn't matter if we implement `refid_check`. What i meant is, instead of saving `held_operations: list[refid]` we could save `held_operations: list[op]`. Because we nether user the refid for anything, we actually want the operations.
If we want to improve the performance, we can make it a `dict[refid, list[op]` to increase the speed of finding the correct entry.

----



I believe, that we do not need to save `held_operations`. It might be enough to save `refid_check` as you already described it. If we find an suitable operation in the loop, we just add it directly instead of saving it into the `held_operations`.

If we want to save only the first. No problem, just add it; it doesn't harm for now to have it added; we can still raise an error later, if the second couldn't be found.

If we want to save the second; just wait for it.

The `refid_check: dict[refid, bool]` dictionary is the key to all that. This dictionary will tell, if we found the first or second item.
```
None: not found (or not existing in the dictionary)
False: one found
True: two found
if we find another entry: raise error, thats not supposed to happen
````

Afterwards check `all(v == True for v in refid_check.values())`. If not all ids could be found twice, raise an error with the conflicting error.

Reply:
Ah, I see. In that case, it might be the easiest way to keep the `held_operations` variable. 

Preprocessing the list (check which ref_ids are duplicated and select the correct operations accordingly, is difficult to implement for different files in our current setup.

Then, we might profit from using a dictionary instead of a list (`dict[ref_id, operations]`)

Reply:
We can add the `held_operations` variable as Book class member (in `__init__`) or as class variable (directly after `class Book:`).